src/Microsoft.Diagnostics.Monitoring/Contracts/IPipeline.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.Contracts
+{
+    /// <summary>
+    /// A pipeline controls data which is flowing from some source to sink asynchronously.
+    /// This interface is allows the flow to be started and stopped. The concrete class
+    /// determines what data is being collected and where it will flow to.
+    /// 
+    /// The pipeline is logically in one of these states:
+    /// Unstarted - After the object is constructed and prior to calling RunAsync or
+    /// StopAsync. No data is flowing.
+    /// Running - The pipeline is doing whatever asynchronous work is necessary to flow
+    /// data. Unstarted transitions to Running with a call to RunAsync()
+    /// Stopping - The pipeline is doing a graceful shutdown to stop receiving any new
+    /// data and drain any in-flight data to the sink. Unstarted or Running transitions to
+    /// Stopping with a call to StopAsync(). Pipelines may also automatically enter a Stopping
+    /// state when there is no data left to receive from the source.
+    /// Stopped - All asynchronous work has ceased and the pipeline can not be restarted. This
+    /// transition happens asynchronously from the stopping state when there is no
+    /// work left to be done. The only way to be certain you have reached this state is to
+    /// observe that the Task returned by StopAsync() or RunAsync() is completed or cancelled,
+    /// usually by awaiting it.
+    /// </summary>
+    public interface IPipeline : IAsyncDisposable
```

Comment:
I'm wondering if it makes sense to refactor the assemblies such that:
- The IPipeline and related types are in a new assembly named Microsoft.Diagnostics.Pipelines
- The event pipe implementations are in a new assembly named Microsoft.Diagnostics.Pipelines.EventPipe
- The IDiagnosticServices feel like a dotnet-monitor implementation detail, so they could be moved to dotnet-monitor.
- After that, all that's left is the IEndpointInfoSource and related classes, which could stay in the Microsoft.Diagnostics.Monitoring assembly or possibly move to Microsoft.Diagnostics.NETCore.Client.

In my mind, the pipeline types don't necessarily have much to do with monitoring but are to facilitate flowing of data and egressing information using the event pipe. Monitoring is an observational process over a period of time.

Reply:
I don't disagree with your points, but I'm not sure we want such high shared assembly granularity. Maybe Microsoft.Diagnostics.Monitoring can be renamed though.

Reply:
@sywhang @josalem The basic design here is:
- There's a shared pipeline assembly that contains IPipeline and some shared implementation code.
- There's a concrete pipeline implementation assembly. Currently this contains pipelines that use TraceEvent.
- There will be other pipelines in the future, (for example we would want to unify dump implementation between dotnet-monitor and dotnet-dump, but not force anyone who wants to use dump collection to take a dependency on TraceEvent).